### ISSUE TITLE
Notification module throws error for orders with several partial payments

### ISSUE THREAD

**Issue by Arkadiusz-Choczaj-Epam** [Adyen/adyen-commercetools#1150]
**Describe the bug**
Error `Cannot read properties of undefined (reading 'custom')` thrown by notification module for orders with several partial payments. 
Happens for orders paid through Adyen web-drop-in component with two payment methods, like i.e. partially with gift card (in our case Givex), and partially with other payment method

**To Reproduce**
(lets assume for example purposes that order is for 100 EUR, 20 EUR is paid with gift card and the other 80 EUR by visa)
What happens:
- payment object created in Commercetools with `amountPlanned` 100 EUR
- payment object updated with `createSessionRequest`
- payment object updated with `createSessionResponse`, extension module also triggers `setKey` update action, and `key` is set to value coming from `merchantReference` at this stage
- notification module receives notification about 1st successful payment: by gift card for 20 EUR, triggers update of `key` to value coming from `pspReference=SAMPLE_PSP_REF_1` received in notification 
- notification module receives notification about 2nd successful payment: by visa for 80 EUR, `merchantReference` is the same as for 1st payment, but `pspReference=SAMPLE_PSP_REF_2` is different now, payment in Commercetools is not found, neither by `pspReference=SAMPLE_PSP_REF_2` nor by `merchantReference` (because it was replaced earlier with 1st payment `pspReference=SAMPLE_PSP_REF_1`)
error is thrown:
```
{
    "notification": [
        {
            "success": "true",
            "eventCode": "AUTHORISATION",
            "pspReference": "SAMPLE_PSP_REF_2",
            "eventDate": "2024-02-21T13:35:03+01:00"
        }
    ],
    "level": 50,
    "msg": "Unexpected exception occurred.",
    "hostname": "***",
    "pid": 1,
    "name": "ctp-adyen-integration-notifications",
    "cause": {
        "name": "TypeError",
        "stack": "TypeError: Cannot read properties of undefined (reading 'custom')\n at Object.processNotification (file:///app/src/handler/notification/notification.handler.js:59:14)\n at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n at async handleNotification (file:///app/src/api/notification/notification.controller.js:27:7)\n at async Server.<anonymous> (file:///app/src/server.js:20:9)",
        "message": "Cannot read properties of undefined (reading 'custom')"
    },
    "v": 0
}
```

Result:
- payment object with `amountPlanned` 100 EUR is updated only with details for 1st 20 EUR payment
- for 2nd payment notification module responds to Adyen webhook with `"notificationResponse" : "[accepted]"`
- Notification module throws error, info about 2nd payment never reaches Commercetools

**Expected behavior**
Payment object(s) in Commercetols updated with all transaction actually done in Adyen

**Additional context**
looks like the root cause is similar as in issue: https://github.com/Adyen/adyen-commercetools/issues/1144

**Comment by logeecom** [comment]
Hey @Arkadiusz-Choczaj-Epam,

Thank you for reaching out, our team is investigating this matter and we will update you once we have more information.

Best regards.

**Comment by logeecom** [comment]
Hey @Arkadiusz-Choczaj-Epam 

Please note that we've configured the payment key to a PSP reference from webhook to facilitate the custom refund reference feature. Keeping the original merchant reference for the payment key would compromise backward compatibility with changes implemented in version [9.10.0](https://github.com/Adyen/adyen-commercetools/releases/tag/v9.10.0). We acknowledge this request for future development and aim to formulate a solution that maintains backward compatibility.

However, you could implement a workaround for this specific scenario, where each Adyen transaction will be assigned to a different commercetools payment. In this way, each partial Adyen transaction will have its commercetools payment instance and webhooks will update the payment instance they belong to based on the original PSP reference value.

If you have any further questions or need clarification, please feel free to reach out.

**Comment by logeecom** [comment]
Hey @Arkadiusz-Choczaj-Epam,

We appreciate your suggestion. After careful consideration, we have decided not to proceed with this request at the moment. However, your suggestion has been noted and added to our internal backlog.

Thank you!
